### PR TITLE
Make Function ChainRec

### PR DESCRIPTION
**Pull request by safareli**
Recently [wrote `ChainRec` instance for Function](https://github.com/fantasyland/fantasy-land/issues/212#issuecomment-267411320) so here it is ✨

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -105,6 +105,24 @@
...
 
+  // type Iteration a = { value :: a, done :: Boolean }
+
+  //  iteratoinNext :: a -> Iteration a
+  function iteratoinNext(x) {
+    return {
+      value: x,
+      done: false
+    };
+  }
```
Comment (davidchambers) [inline index.js]: We have a typo: `:%s/iteratoinNext/iterationNext/g`. Also, I'd prefer more compact formatting:

```javascript
  function iterationNext(x) { return {value: x, done: false}; }
```

File: index.js [hunk 2]
```diff
@@ -738,6 +754,17 @@
     return function(_) { return x; };
   }
 
+  //  Function$chainRec :: ((a -> c, b -> c, a) -> (z -> c), a) -> (z -> b)
+  function Function$chainRec(f, i) {
```
Comment (davidchambers) [inline index.js]: Let's rename the second parameter `x` to match `Array$chainRec`. :)

File: test/index.js [hunk 3]
```diff
@@ -678,6 +678,19 @@ test('chainRec', function() {
...
+      function(next, done, n) {
+        return n === 30000
+          ? Z.map(done, function(env) { return n + env.inc; })
+          : Z.map(next, function(env) { return n + env.step; });
+      },
+      0
+    )({step: 2, inc: 100}),
+    30100
+  );
+
```
Comment (davidchambers) [inline test/index.js]: I'd like to name the anonymous function (as we did in the case of `squash` below) so we can write:

```javascript
  eq(Z.chainRec(Function, TK, 0)({step: 2, inc: 100}), 30100);
```

`TK` is a placeholder name. I've no idea what the function should be named. ;)

File: index.js [hunk 4]
```diff
@@ -105,6 +105,24 @@
     };
   }
 
+  // type Iteration a = { value :: a, done :: Boolean }
```
Comment (davidchambers) [inline index.js]: Two spaces after `//`, please. :)

File: test/index.js (hunk 3 above)
Comment (safareli) [inline test/index.js]: `stepper`?

**Comment by safareli** [comment]
fixed ✅

**Comment by davidchambers** [comment]
Looks great! The final change I'd like you to make is to move the test for `Function.chainRec` after the tests for `Array.chainRec` (to match the order in which the definitions appear in __index.js__).

Please also squash the commits. :)

**Comment by safareli** [comment]
🚀

**Comment by davidchambers** [comment]
:deciduous_tree:

**Pull request merged**